### ISSUE TITLE
Allow passing the TeX file name to the generic compiler?

### ISSUE THREAD

**Issue by user202729** [lervag/vimtex#2619]
**Is your feature request related to a problem? Please describe it.**

As I can see currently the generic compiler only allow specifying a command to be executed, but the command does not have access to the information which TeX file is the current file.

**Describe the solution you'd like**
Allow specifying it somehow. (perhaps with place holders such as `%f`? But that way `%f` itself cannot be part of a command)

**Describe alternatives you've considered**
Currently I use `$(ls -t *.tex | head -1)` as a hack.

**Comment by lervag** [comment]
Good idea. I image this means to allow place holders in the `g:vimtex_compiler_generic.command`. The current documentation is:

https://github.com/lervag/vimtex/blob/bd0173e7d3c0097c5b855265ce83f5863f073c85/doc/vimtex.txt#L1198-L1200

My first idea of the change would be this:

```
  command~
    This is the command to run to start compilation. This can be any command,
    and the command is run from the project root. It supports the following
    command substitutions:

      - `%r`: Path to project root
      - `%f`: Path to project main file

    To avoid the substitution, use tw `%`s; e.g. `%%r` is parsed as `%r` verbatim.
```

Let me know what you think. Did I misunderstand; should we include more/other substitutions?

**Comment by user202729** [comment]
Looks good. (there's a typo ``tw `%`​`` by the way)

Nitpick, should `%x` (where `x` is not a supported substitution) be written as `%x` or `%%x`? (the latter would be more consistent)

The user should use `'` around the substitution, of course this won't support the file name containing `'` but it's likely not a high priority anyway.

**Comment by lervag** [comment]
> Nitpick, should `%x` (where `x` is not a supported substitution) be written as `%x` or `%%x`? (the latter would be more consistent)

The latter; I'll probably implement it by
1. substituting `%%` with another symbol, e.g. `¤`, then
2. substituting the recognized patterns, e.g. `%f` to `/path/to/project/main.tex`, finally
3. reversing `¤` to `%`.

> The user should use `'` around the substitution, of course this won't support the file name containing `'` but it's likely not a high priority anyway.

This is unclear to me. I believe the user would write e.g.

```vim
let g:vimtex_compiler_generic = {
      \ 'command' : 'command %f %r',
      \}
```

Which would substitute to `command /path/to/project/main.tex /path/to/project`. I believe we should avoid escaping spaces and similar and instead allow the user to surround the expansions with double quotes if they need to `"%f"`. An alternative would be to automatically add the double quotes or to escape spaces and similar in front. I'm not sure if that is the most expected behaviour, though.

**Comment by lervag** [comment]
Are you still interested in this? I'm waiting for a reply to my previous comment here...

**Comment by user202729** [comment]
I since then have switched to something else and don't need this feature anymore -- nevertheless, my point is your idea would cause issue if the file name contain either `"` or `'`.

**Comment by lervag** [comment]
> I since then have switched to something else and don't need this feature anymore

That's fine. I'm curious, what have you switched to?

> nevertheless, my point is your idea would cause issue if the file name contain either `"` or `'`.

Yes, perhaps. But I think the standard `shellescape()` would fix that (and similar things).

As I don't have so much time these days, and since you are no longer interested in this, I'll close this issue as "won't fix".

**Comment by user202729** [comment]
Alright, no problem.

I realize it's difficult to use "continuous compilation" feature using generic compiler anyway, so I make my script "emulate" latexmk's front end then modify `vimtex_compiler_latexmk["executable"]`: https://github.com/user202729/tex-fast-recompile#latexmk-emulation-mode

**Comment by lervag** [comment]
Cool, thanks for sharing. By the way, you don't need this line:

```vim
let g:vimtex_compiler_method='latexmk'
```

Because `latexmk` is already the default. Thus, you can remove [line 50](https://github.com/user202729/tex-fast-recompile/blob/main/README.md?plain=1#L50) in README.md.

I'm curious: do you have any benchmarks for how much faster compilation speeds you get with this compared to standard latexmk/pdflatex?

**Comment by user202729** [comment]
Currently I don't have rigorous benchmark, but rough estimation is "however long it takes to compile an empty document using the current preamble", so it would be something like

* negligible with pdflatex/empty preamble
* 0.5s for lualatex
* 1-1.5s for tikz/unicode-math/something similarly heavy

Of course, for documents with long body the constant-offset gain is negligible compared to the time to compile the body, in that case standard tricks are needed (includeonly, comment out unimportant parts etc.)

**Comment by lervag** [comment]
Thanks! :)
